lib/charms/mysql/v0/mysql.py

```diff
@@ -825,13 +857,66 @@ def get_cluster_primary_address(self, connect_instance_address: str = None) -> s
 
         return matches.group(1)
 
+    def get_cluster_members_addresses(self) -> Optional[Iterable[str]]:
+        """Get the addresses of the cluster's members.
+
+        Returns:
+            Iterable of members addresses
+        """
+        get_cluster_members_commands = (
+            f"shell.connect('{self.cluster_admin_user}:{self.cluster_admin_password}@{self.instance_address}')",
+            f"cluster = dba.get_cluster('{self.cluster_name}')",
+            "members = ','.join((member['address'] for member in cluster.describe()['defaultReplicaSet']['topology']))",
+            "print(f'<MEMBERS>{members}</MEMBERS>')",
```

Comment:
Perhaps I'm not understanding MySQL, but I'm confused as to what's going on here.
Are we exec-ing generated Python code against `/usr/bin/mysqlsh`?

If so that seems quite dangerous and REALLY hard to debug.

Is something stopping us from doing the same as a subprocess using returned data from MySQL API to get the values we want (I think 'address','members' etc)?

Reply:
Feel free to ignore this if this is totally fine and safe. Just looks a bit wild 😄

Reply:
The thing is, `mysqlsh` (which is an advanced command line tool) has an extensive (python) [API](https://dev.mysql.com/doc/dev/mysqlsh-api-python/8.0/annotated.html) to manage the cluster and get all kinds of metadata on the cluster. If we were to use a mysql client, these utilities wouldn't be available to us. `mysqlsh` has the ability to interpret and execute python code, which is what we're doing in a subprocess (by calling `_run_mysqlsh_script` -> we write the python code to a file and then pass the file to `mysqlsh` to execute)